chain/network/src/peer/peer_actor.rs

```diff
@@ -938,39 +953,42 @@ impl PeerActor {
-                                .values()
-                                // Do not send the data back.
-                                .filter(|p| peer_id != p.peer_info.id)
-                                .map(|p| p.send_accounts_data(new_data.clone()))
-                                .collect();
-                            futures_util::future::join_all(handles).await;
-                        }
-                        err.map(|err| match err {
-                            accounts_data::Error::InvalidSignature => {
-                                ReasonForBan::InvalidSignature
+                    wrap_future(
+                        async move {
+                            // Early exit, if there is no data in the message.
+                            if msg.accounts_data.is_empty() {
+                                return None;
                             }
-                            accounts_data::Error::DataTooLarge => ReasonForBan::Abusive,
-                            accounts_data::Error::SingleAccountMultipleData => {
-                                ReasonForBan::Abusive
+                            // Verify and add the new data to the internal state.
+                            let (new_data, err) =
+                                pms.accounts_data.clone().insert(msg.accounts_data).await;
+                            // Broadcast any new data we have found, even in presence of an error.
+                            // This will prevent a malicious peer from forcing us to re-verify valid
+                            // datasets. See accounts_data::Cache documentation for details.
+                            if new_data.len() > 0 {
+                                let handles: Vec<_> = pms
+                                    .tier2
+                                    .load()
+                                    .ready
+                                    .values()
+                                    // Do not send the data back.
+                                    .filter(|p| peer_id != p.peer_info.id)
+                                    .map(|p| p.send_accounts_data(new_data.clone()))
+                                    .collect();
+                                futures_util::future::join_all(handles).await;
                             }
-                        })
-                    })
+                            err.map(|err| match err {
+                                accounts_data::Error::InvalidSignature => {
+                                    ReasonForBan::InvalidSignature
+                                }
+                                accounts_data::Error::DataTooLarge => ReasonForBan::Abusive,
+                                accounts_data::Error::SingleAccountMultipleData => {
+                                    ReasonForBan::Abusive
+                                }
+                            })
+                        }
+                        .in_current_span(),
```

Comment:
NB: putting the `in_current_span` here means that the code within `map` is still not getting instrumented in any way.

I imagine you can’t easily change this to `wrap_future(...).map(...).in_current_span()` because `ActixFuture` isn’t a proper future? That’s quite unfortunate.

We need to keep in mind to never do anything substantial in these actor combinators.

Reply:
> because ActixFuture isn’t a proper future?

Correct

> to never do anything substantial in these actor combinators.

so far I haven't seen anything substantial in these combinators

Reply:
> so far I haven't seen anything substantial in these combinators

Right.

Consider a scenario: somebody comes along in the future and adds something more to one of these `map`s. This new addition happens to do something that takes a long time. None of that time shows up in any of the traces during development. Knowing that the modified code was instrumented, they are confident in shipping the changes out to production, where everything goes :boom:. I’d argue that in such a scenario lack of the resilience in the instrumentation would easily be accountable for allowing the :boom: to occur at all.

Now, it isn’t something to block this PR over. However, I do think it is a good idea to at least file an issue and perhaps give some resources towards figuring out a solution for this eventually.